### PR TITLE
Recover from impl Trait in type param bound

### LINKED ISSUE #102182: Recover from `impl Trait` in type param bound
When writing `fn foo<T: impl Trait>() {}` the current error is "expected one of these tokens", without any attempt to recover the parse. It should instead be closer to:

```
error[E0404]: expected trait, found struct `S`
 --> src/lib.rs:2:11
  |
2 | fn foo<T: S>() {}
  |           ^ not a trait
```

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #102182
r? @estebank

### CODE REVIEW

File: compiler/rustc_parse/src/parser/generics.rs [hunk 1]
```diff
@@ -31,13 +33,33 @@ impl<'a> Parser<'a> {
...
+            // recover from `impl Trait` in type param bound
+            if self.token.is_keyword(kw::Impl) {
+                let impl_span = self.token.span;
+                let snapshot = self.create_snapshot_for_diagnostic();
+                match self.parse_ty() {
+                    Ok(p) => {
+                        if let TyKind::ImplTrait(..) = (*p).kind {
+                            let span = impl_span.to(self.token.span.shrink_to_lo());
+                            let mut err =
+                                self.struct_span_err(span, "expect trait, found `impl trait` type");
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/generics.rs:45]:
```suggestion
                                self.struct_span_err(span, "expected trait bound, found `impl Trait` type");
```

File: compiler/rustc_parse/src/parser/generics.rs [hunk 2]
```diff
@@ -31,13 +33,33 @@ impl<'a> Parser<'a> {
         let mut colon_span = None;
         let bounds = if self.eat(&token::Colon) {
             colon_span = Some(self.prev_token.span);
+            // recover from `impl Trait` in type param bound
+            if self.token.is_keyword(kw::Impl) {
+                let impl_span = self.token.span;
+                let snapshot = self.create_snapshot_for_diagnostic();
+                match self.parse_ty() {
+                    Ok(p) => {
+                        if let TyKind::ImplTrait(..) = (*p).kind {
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/generics.rs:42]:
```suggestion
                        if let TyKind::ImplTrait(_, bounds) = (*p).kind {
```

File: compiler/rustc_parse/src/parser/generics.rs [hunk 3]
```diff
@@ -31,13 +33,33 @@ impl<'a> Parser<'a> {
...
+                            let span = impl_span.to(self.token.span.shrink_to_lo());
+                            let mut err =
+                                self.struct_span_err(span, "expect trait, found `impl trait` type");
+                            err.span_label(span, "not a trait");
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/generics.rs:46]:
```suggestion
                            err.span_label(span, "not a trait");
                            if let [bound, ..] = &bounds[..] {
                                err.span_suggestion_verbose(
                                    impl_span.until(bound.span()),
                                    "use the trait bounds directly",
                                    String::new(),
                                    Applicability::MachineApplicable,
                                );
                            }
```

**Review by estebank — APPROVED** [review]
Minor nitpicks

File: compiler/rustc_parse/src/parser/generics.rs (hunk 3 above)
Comment (chenyukang) [reply compiler/rustc_parse/src/parser/generics.rs:48]: Updated, thanks!

**Pull request merged**